In [1]:
import pandas as pd
import sqlite3

In [2]:
albums_query = """
SELECT *
FROM albums
"""
with sqlite3.connect('data/chinook.db') as chin_con:
    albums_df = pd.read_sql(albums_query, chin_con)

albums_df.head()

AlbumId                                  Title  ArtistId
0        1  For Those About To Rock We Salute You         1
1        2                      Balls to the Wall         2
2        3                      Restless and Wild         2
3        4                      Let There Be Rock         1
4        5                               Big Ones         3

In [3]:
albums_df.shape

(347, 3)

### Question (a)
There are 347 albums in the database.

In [4]:
q_b= """
SELECT COUNT(DISTINCT ArtistId) AS unique_artists
FROM albums
"""
with sqlite3.connect('data/chinook.db') as con:
    b_df = pd.read_sql(q_b, con)
b_df

unique_artists
0             204

### Question (b)
There are 204 unique artists in the albums table.

In [6]:
q_c = """
SELECT COUNT(*) AS media_type_count
FROM media_types
"""
with sqlite3.connect('data/chinook.db') as con:
    c_df = pd.read_sql(q_c, con)
c_df

media_type_count
0                 5

### Question (c)
There are 5 different media types.

In [7]:
q_d = """
SELECT
    mt.MediaTypeId,
    mt.Name AS media_type,
    COUNT(t.TrackId) AS track_count
FROM media_types mt
LEFT JOIN tracks t
    ON t.MediaTypeId = mt.MediaTypeId
GROUP BY mt.MediaTypeId, mt.Name
ORDER BY track_count ASC
"""
with sqlite3.connect('data/chinook.db') as con:
    d_df = pd.read_sql(q_d, con)
d_df

MediaTypeId                   media_type  track_count
0            4     Purchased AAC audio file            7
1            5               AAC audio file           11
2            3  Protected MPEG-4 video file          214
3            2     Protected AAC audio file          237
4            1              MPEG audio file         3034

### Question (d)
Yes, all media types are represented in the tracks table.

In [8]:
q_e = """
SELECT
    mt.Name AS media_type,
    COUNT(t.TrackId) AS track_count
FROM tracks t
JOIN media_types mt
    ON t.MediaTypeId = mt.MediaTypeId
GROUP BY mt.Name
ORDER BY track_count DESC
"""
with sqlite3.connect('data/chinook.db') as con:
    e_df = pd.read_sql(q_e, con)
e_df

media_type  track_count
0              MPEG audio file         3034
1     Protected AAC audio file          237
2  Protected MPEG-4 video file          214
3               AAC audio file           11
4     Purchased AAC audio file            7

### Question (e)
The table above shows the number of tracks for each media type.

In [9]:
q_f = """
SELECT
    City,
    COUNT(*) AS customer_count
FROM customers
GROUP BY City
ORDER BY customer_count DESC
LIMIT 1
"""
with sqlite3.connect('data/chinook.db') as con:
    f_df = pd.read_sql(q_f, con)
f_df

City  customer_count
0  São Paulo               2

### Question (f)
The city with the largest number of customers is Sao Paulo.

In [11]:
q_g = """
SELECT
    g.Name AS genre,
    COUNT(t.TrackId) AS track_count
FROM tracks t
JOIN genres g
    ON t.GenreId = g.GenreId
GROUP BY g.Name
ORDER BY track_count DESC
LIMIT 1
"""
with sqlite3.connect('data/chinook.db') as con:
    g_df = pd.read_sql(q_g, con)
g_df

genre  track_count
0  Rock         1297

### Question (g)
The most common genre of tracks is Rock.